### ISSUE TITLE
Elaborate on the use cases for a Digest

### ISSUE THREAD

**Issue by jyasskin** [ioggstream/draft-polli-resource-digests-http#15]
https://ioggstream.github.io/draft-polli-resource-digests-http/draft-polli-resource-digests-http.html#goals says a goal is "Digest coverage for either the resource’s representation data or selected representation data communicated via HTTP.", but it's not clear to me why a client and server would use a digest. Some options include:

* To detect message corruption. This is the only reason I can think of to use the ADLER32, CRC32c, MD5, SHA, UNIXsum, or UNIXcksum algorithms, but I think it's unnecessary with modern transport protocols?
* To support signatures, for example in [Signed Exchanges](https://github.com/WICG/webpackage).
* To give the client a hint about which SRI hash will be used, to maybe help with https://github.com/w3c/preload/issues/127.

Are there others? If those are the only ones, and message corruption isn't needed anymore, we could deprecate all the non-cryptographic digest algorithms.

**Comment by ioggstream** [comment]
Hi @jyasskin!

## Digest use-cases

The reason to use a digest is detailed in [introduction](https://ioggstream.github.io/draft-polli-resource-digests-http/draft-polli-resource-digests-http.html#introduction) and mentions SRI.

If I add the Signature use-case would it suffice?

## Deprecate non-crypto algorithms

I agree. I was just not sure whether deprecating that algorithms could limit the adoption of this draft, so I decided to ask the ietf. issue moved in #16.

**Comment by LPardue** [comment]
One use of digests is to ensure end-to-end integrity over multiple hops. While each hop has its own transport-layer integrity, there is a possibility that the true contents get somehow malformed (e.g. broken at rest, cosmic ray bit flips, buggy compressor, buggy manipulation etc.). Having an ability to validate integrity along the end-to-end path can help to reduce end-user errors. This has some nice side effect in certain deploments. One might imagine that intermediaries can detect errors from further up the chain and mitigate the problem before fanning out a known-bad resource to a large client population.

Another benefit is piecing together a resource from different sources (e.g. different servers that perhaps apply different content codings. The identiy digest allows the "final detination" to detect that the application-layer tasks completed properly, before handing off to say the HTML parser, video player etc.

For these reasons, I think that some use cases may want, and would benefit from, algorithms like MD5 etc. But I'm open to deprecating algorithms if there are good reasons. I tried to find some kind of official IETF document that makes recommendations in this space but came up empty handed.

**Comment by jyasskin** [comment]
I'm nervous about the end-to-end integrity use case for Digest: it only protects the representation data, while the headers are important for making sure the application interprets that data correctly. Buggy manipulation seems as likely to swap out a content-type, which wouldn't be caught by Digest, as to change the representation data, which would.

That said, if folks are actually using Digest to detect bugs, it makes sense to document what they're doing in this RFC.

**Comment by LPardue** [comment]
These are good points, thanks for writing them up.

I agree on its own that Digest has some limited guarantees, but the spec does try to highlight them. 

The defenses here are mainly to avoid accidental or insipid errors, rather than malice. Defense in depth is required for that (using TLS for example). 

Combining Digest with a Signature header (yours or Cavages) is one way to enhance the protections.

If we need to spell out things better in the spec, I think that's reasonable.

**Comment by ioggstream** [comment]
@jyasskin 
> [...] Digest: it only protects the representation data, while the headers are important

True, this is highlighted [here](https://github.com/ioggstream/draft-polli-resource-digests-http/blob/master/draft-polli-resource-digests-http.md#usage-in-signatures) but I filed #20 to move this discussion to  a proper thread.

**Comment by jyasskin** [comment]
Note that end-to-end integrity against accidental corruption or errors (i.e. not attackers) could be ensured via a checksum-Digest that covers both the headers and the body. Only defense against attackers requires a signature or MAC by a trusted key.

**Comment by ioggstream** [comment]
I'm addressing those concerns here:

- https://github.com/ioggstream/draft-polli-resource-digests-http/pull/21/files
- https://github.com/ioggstream/draft-polli-resource-digests-http/commit/3e906082b4f6726ca0882f5f3ced0c52d371f15b [explicit `representation data`]

Feel free to emend!

**Comment by LPardue** [comment]
@jyasskin did we fulfilled your initial request for Digest use cases?
